docs/markdown/Reference-manual.md

```diff
@@ -326,8 +326,20 @@ false otherwise.
 ```
 
 Create a custom top level build target. The only positional argument
-is the name of this target and the keyword arguments are the
-following.
+is the name of this target and cannot contain path separators (`/` or `\`).
+The name of custom target might not be used by every backends, for instance with
+the Ninja backend, `subdir/meson.build` containing the example below,
+`ninja -C builddir foo` or `ninja -C builddir subdir/foo` won't work,
+it is instead `ninja -C builddir subdir/file.txt`. Howerver, `meson compile subdir/foo`
+is accepted.
```

Comment:
> After more digging, it is indeed a bad idea to have path separators in the custom target name, because it is used in the introspection data, it breaks meson compile <target> command.

I'm *very* surprised that `meson compile` accepts nonsense targets that `ninja` doesn't accept, and I hate it.

Apparently, `meson compile` only *usually* does more or less the same thing as ninja because target names are usually manually coded as the output basename based on superstition. e.g.

```
ninja doc/pacman.8
meson compile doc/pacman.8

ninja scripts/po/tr/LC_MESSAGES/pacman-scripts.mo
```

all work, but 
```
meson compile ninja scripts/po/tr/LC_MESSAGES/pacman-scripts.mo
``` 
reports an error and only using the display name works.

I am now firmly convinced that `meson compile` is completely evil and must be immediately removed.

Reply:
Each backend has a different way to identify targets, ninja uses exclusively output files path. It makes `meson compile` even more important to have common target names across backends. But in any case, that's offtopic.

As far as this PR is concerned, I think it does make sense to forbid path separator in custom_target() name, because that's consistent with all other targets. Prepending the subdir in ninja description automatically also makes a lot of sense because that's exactly what it does for other targets already: `executable('app', 'foo.c')` in subdir `foo` produce
```
[1/2] Compiling C object foo/app.p/foo.c.o
[2/2] Linking target foo/app
```

Reply:
@eli-schwartz any objection on merging this now?

Reply:
The PR itself seems fine, I'm just being my usual grumpish self about the existence of mcompile. :)

Offtopic: if we *actually* wanted to support this sort of "common target name", the appropriate solution would have been to add .PHONY targets to the meson.build, but apparently the time for that has passed...